DESIGN.md

```diff
@@ -0,0 +1,471 @@
+  - Generate a unique job ID
+  - Create a cgroup for the job and apply resource limits
+  - Create a new process using `syscall.ForkExec()` with the following attributes:
+    - `ProcAttr.Env` = `[]string{safePATH}` (set a safe PATH for the job)
+    - `SysProcAttr.Setpgid` = true (start as its own process group leader).
+    - `SysProcAttr.Pdeathsig` = SIGKILL (if the supervisor dies, the child is killed).
+    - `SysProcAttr.Credential` = {Uid,Gid} set to `nobody` for normal users or `1000:1000` for `admin`.
+    - Set up stdout and stderr to write to `/var/lib/jobs/<jobid>/stdout.log` and `/var/lib/jobs/<jobid>/stderr.log`
+      - ProcAttr.Files = []uintptr{stdinFD, stdoutFD, stderrFD} mapping FD 0/1/2 to:
+        - stdin: `/dev/null`
+        - stdout: `<jobdir>/stdout.log`
+        - stderr: `<jobdir>/stderr.log`
+  - Immediately move the process to the cgroup by writing its PID to `/sys/fs/cgroup/jobs/<jobid>/cgroup.procs`
+  - Wait for the processes to finish, capturing its exit code
+
+```go
+	attr := &syscall.ProcAttr{
+		Env:   []string{safePATH},
+		Files: []uintptr{devNull.Fd(), stdout.Fd(), stderr.Fd()},
+		Sys: &syscall.SysProcAttr{
+			Setpgid:    true, // Start as its own process group leader
+			Pdeathsig:  syscall.SIGKILL, // If the supervisor dies, the child is killed
+            // Set the user and group to `nobody` for normal users or `1000:1000` for admin users
+			Credential: &syscall.Credential{Uid: uid, Gid: gid},
+		},
+	}
+
+	argv := append([]string{fullPath}, cmdArgs...)
+	pid, err := syscall.ForkExec(fullPath, argv, attr)
+    // argv is the command and its arguments
+    // if the command is `ls` then we will need to use `exec.LookPath("ls")` to get the full path to the executable
+    // not shown here for brevity
+```
+
+##### Security note about Start()
+- The **only** environment variable that will be set is `PATH` to a safe value. This is to prevent any potential PATH injection attacks. The `PATH` will be set to a safe value like `/usr/bin:/bin` or similar.
+
+#### Key Subcomponents: (other potential packages)
+- `devinfo`: A helper to provide device info. Like what /dev to set the IO cgroup limits on. This will be determined by grabbing the device where the root (/) dir is mounted.  (ie. /dev/sda might be 8:0)
+- `cgroups`: Actually creates and deletes cgroups v2 limits (implementation described below)
+- `resources`: A helper package to abstract away all the many low level resource requirements (TBD if needed)
+
+#### Cgroups
+
+##### Notes
+- Cgroups v2 will be used. Kernel version 5.2+ is required for this functionality
+- Upon job termination the Job Controller via `joblib` will cleanup the cgroup files
+
+#### Details
+1. Create a custom cgroup directory `/sys/fs/cgroup/jobs/<jobid>`
+  - Create `/sys/fs/cgroup/jobs` if it doesnt exist yet
+2. Inside  `<jobid>` directory:
+  - set `memory.max` in bytes (one token)
+  - set `cpu.max` to "<quota> <period>" where they are in microseconds. (2 tokens)
```

Comment:
How are you going to get `<quota>` and `<period>` if the user provided a single string like `500m`?